drivers/i2c/i2c_esp32.c

```diff
@@ -0,0 +1,781 @@
+/*
+ * Copyright (c) 2017 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/* Include esp-idf headers first to avoid redefining BIT() macro */
+#include <soc/dport_reg.h>
+#include <soc/i2c_reg.h>
+#include <rom/gpio.h>
+#include <rom/ets_sys.h>
+#include <soc/gpio_sig_map.h>
+
+#include <errno.h>
+#include <gpio.h>
+#include <i2c.h>
+#include <misc/util.h>
+#include <string.h>
+
+#define I2C_ESP32_NUM_CMDS 16
+#define I2C_ESP32_BUFFER_SIZE 32
+#define I2C_ESP32_TIMEOUT_MS 100
+#define I2C_ESP32_SPIN_THRESHOLD 10
+#define I2C_ESP32_YIELD_THRESHOLD (I2C_ESP32_SPIN_THRESHOLD / 2)
+#define I2C_ESP32_TIMEOUT \
+	((I2C_ESP32_YIELD_THRESHOLD) + (I2C_ESP32_SPIN_THRESHOLD))
+
+enum i2c_esp32_opcodes {
+	I2C_ESP32_OP_RSTART,
+	I2C_ESP32_OP_WRITE,
+	I2C_ESP32_OP_READ,
+	I2C_ESP32_OP_STOP,
+	I2C_ESP32_OP_END
+};
+
+struct i2c_esp32_cmd {
+	u32_t num_bytes : 8;
+	u32_t ack_en : 1;
+	u32_t ack_exp : 1;
+	u32_t ack_val : 1;
+	u32_t opcode : 3;
+	u32_t : 17;
+	u32_t done : 1;
+};
+
+struct i2c_esp32_data {
+	union dev_config config;
+	u16_t address;
+
+	struct k_sem fifo_sem;
+	struct k_sem transfer_sem;
+};
+
+struct i2c_esp32_config {
+	const struct {
+		volatile u32_t *ctr;
+		volatile u32_t *fifo_cfg;
+		volatile u32_t *fifo;
+		volatile u32_t *slave_addr;
+		volatile u32_t *intr;
+		volatile u32_t *intr_clr;
+		volatile u32_t *intr_status;
+		volatile u32_t *intr_raw;
+		volatile u32_t *status;
+		volatile u32_t *cmd_base;
+		volatile u32_t *timeout;
+		volatile u32_t *scl_low_period;
+		volatile u32_t *scl_high_period;
+		volatile u32_t *scl_start_h_time;
+		volatile u32_t *scl_rstart_s_time;
+		volatile u32_t *scl_stop_h_time;
+		volatile u32_t *scl_stop_s_time;
+		volatile u32_t *sda_hold_time;
+		volatile u32_t *sda_sample_time;
+	} regs;
+
+	const struct {
+		int sda_out;
+		int sda_in;
+		int scl_out;
+		int scl_in;
+	} sig;
+
+	const struct {
+		int scl;
+		int sda;
+		bool scl_pull_up;
+		bool sda_pull_up;
+		bool default_level;
+	} pins;
+
+	const struct {
+		int clk;
+		int rst;
+	} enable_bits;
+
+	const struct {
+		bool tx_lsb_first;
+		bool rx_lsb_first;
+	} mode;
+
+	const struct {
+		int source;
+		int line;
+	} irq;
+
+	const union dev_config default_config;
+};
+
```

Comment:
For all the above declarations, it's usually nicer to move them in a dedicated header file, where you can document them as well.